src/HamiltonianParameters/HamiltonianParameters.cpp

```diff
@@ -272,5 +272,102 @@ GQCP::TwoElectronOperator HamiltonianParameters::calculateSuperGeneralizedFockMa
 };
 
 
+/**
+ *  Constrains and copies the Hamiltonian parameters
+ *
+ *  @param one_op   constraining one electron operator
+ *  @param two_op   constraining two electron operator
+ *  @param lambda   lagrangian multiplier for the constraint
+ *
+ *  @return the constrained Hamiltonian parameters
+ */
+HamiltonianParameters HamiltonianParameters::constrain(const GQCP::OneElectronOperator& one_op, const GQCP::TwoElectronOperator& two_op, double lambda) const {
+
+    OneElectronOperator hc (this->get_h().get_matrix_representation() - lambda*one_op.get_matrix_representation());
+    TwoElectronOperator gc (this->get_g().get_matrix_representation() - lambda*two_op.get_matrix_representation());
+
+    return HamiltonianParameters(this->ao_basis, this->S, hc, gc, this->C);
+}
+
+
+/**
+ *  Constrains and copies the Hamiltonian parameters
+ *
+ *  @param one_op   constraining one electron operator
+ *  @param lambda   lagrangian multiplier for the constraint
+ *
+ *  @return the constrained Hamiltonian parameters
+ */
+HamiltonianParameters HamiltonianParameters::constrain(const GQCP::OneElectronOperator& one_op, double lambda) const {
+
+    OneElectronOperator hc (this->get_h().get_matrix_representation() - lambda*one_op.get_matrix_representation());
+
+    return HamiltonianParameters(this->ao_basis, this->S, hc, this->g, this->C);
+}
+
+
+/**
+ *  Constrains and copies the Hamiltonian parameters
+ *
+ *  @param two_op   constraining two electron operator
+ *  @param lambda   lagrangian multiplier for the constraint
+ *
+ *  @return the constrained Hamiltonian parameters
+ */
+HamiltonianParameters HamiltonianParameters::constrain(const GQCP::TwoElectronOperator& two_op, double lambda) const {
+
+    TwoElectronOperator gc (this->get_g().get_matrix_representation() - lambda*two_op.get_matrix_representation());
+
+    return HamiltonianParameters(this->ao_basis, this->S, this->h, gc, this->C);
+}
+
+
+/**
+ *  Calculates the Mulliken operator for Hamiltonian parameters and a set of GTOs indexes
+ *
+ *  @param gto_list     indexes of the original GTOs on which the Mulliken populations are dependant
+ *
+ *  @return the Mulliken operator for a set of GTOs
+ */
+OneElectronOperator HamiltonianParameters::calculateMullikenOperator(const Vectoru& gto_list) {
+
+
+    if (!this->get_ao_basis()) {
+        throw std::invalid_argument("The Hamiltonian parameters has no underlying GTO basis, Mulliken analysis is not possible.");
+    }
+
+    if (gto_list.size() > this->K) {
+        throw std::invalid_argument("To many GTOs are selected");
+    }
+
+    Eigen::MatrixXd p_a = Eigen::MatrixXd::Zero(this->K, this->K);
+
+    for (size_t index : gto_list) {
```

Comment:
The argument checking feels like it can be circumvented by implementing an object that represents a subset of atomic orbitals.